velox/functions/sparksql/aggregates/FirstLastAggregate.cpp

```diff
@@ -0,0 +1,353 @@
+    if (numeric) {
+      return sizeof(TDataType);
+    } else {
+      return sizeof(SingleValueAccumulator);
+    }
+  }
+
+  void initializeNewGroups(
+      char** groups,
+      folly::Range<const vector_size_t*> indices) override {
+    exec::Aggregate::setAllNulls(groups, indices);
+    if (!numeric) {
+      for (auto i : indices) {
+        new (groups[i] + exec::Aggregate::offset_) SingleValueAccumulator();
+      }
+    }
+  }
+
+  void extractValues(char** groups, int32_t numGroups, VectorPtr* result)
+      override {
+    if (numeric) {
+      BaseAggregate::doExtractValues(
+          groups, numGroups, result, [&](char* group) {
+            return *exec::Aggregate::value<TDataType>(group);
+          });
+    } else {
+      VELOX_CHECK(result);
+      (*result)->resize(numGroups);
+
+      auto* rawNulls = exec::Aggregate::getRawNulls(result->get());
+
+      for (auto i = 0; i < numGroups; ++i) {
+        char* group = groups[i];
+        if (exec::Aggregate::isNull(group)) {
+          (*result)->setNull(i, true);
+        } else {
+          exec::Aggregate::clearNull(rawNulls, i);
+          auto accumulator =
+              exec::Aggregate::value<SingleValueAccumulator>(group);
+          accumulator->read(*result, i);
+        }
+      }
+    }
+  }
+
+  void extractAccumulators(char** groups, int32_t numGroups, VectorPtr* result)
+      override {
+    extractValues(groups, numGroups, result);
+  }
+
+ protected:
+  void updateValue(vector_size_t i, char* group, DecodedVector& decoded) {
+    exec::Aggregate::clearNull(group);
+
+    if (numeric) {
+      auto value = decoded.valueAt<TDataType>(i);
+      *exec::Aggregate::value<TDataType>(group) = value;
+    } else {
+      const auto* indices = decoded.indices();
+      const auto* baseVector = decoded.base();
+      auto* accumulator = exec::Aggregate::value<SingleValueAccumulator>(group);
+      accumulator->write(baseVector, indices[i], exec::Aggregate::allocator_);
+    }
+  }
+};
+
+template <bool numeric, bool ignoreNull, typename T>
+class FirstAggregate : public FirstLastAggregateBase<numeric, T> {
+ public:
+  explicit FirstAggregate(TypePtr resultType)
+      : FirstLastAggregateBase<numeric, T>(resultType) {}
+
+  void addRawInput(
+      char** groups,
+      const SelectivityVector& rows,
+      const std::vector<VectorPtr>& args,
+      bool /* mayPushdown */) override {
+    DecodedVector decoded(*args[0], rows);
+
+    std::vector<vector_size_t> selectedIndexes;
+    selectedIndexes.reserve(rows.countSelected());
+    rows.applyToSelected(
+        [&selectedIndexes](vector_size_t i) { selectedIndexes.push_back(i); });
+
+    if (decoded.isConstantMapping()) {
+      if (!decoded.isNullAt(0)) {
+        rows.applyToSelected(
+            [&](vector_size_t i) { this->updateValue(i, groups[i], decoded); });
+      }
+    } else if (decoded.mayHaveNulls()) {
+      for (auto it = selectedIndexes.crbegin(); it < selectedIndexes.crend();
```

Comment:
Any reason not to use rows.applyToSelected ?

Reply:
`rows.applyToSelected` loop all valid rows in normal order, in first aggregate function, we need loop all valid rows in reversed order, cause first row of group is we need.

I try to add a new method of `rows.applyToSelectedReverse`, but I have no ideas about how to implement it with `bits::forEachSetBit`, so I collect all valid row index then loop with reverse order.

Reply:
I'm not sure this is correct as 'rows' contains different groups. 

Why can't we use rows.applyToSelected and implement updateValue for 'first' to keep the first value, while for 'last' to keep the last value?

Reply:
> I'm not sure this is correct as 'rows' contains different groups.

I have verified this logic, it's correct.

> Why can't we use rows.applyToSelected and implement updateValue for 'first' to keep the first value, while for 'last' to keep the last value?

This way is correct in global aggregation which all rows in one group, the rows are deterministic and we only maintain one value.
But for aggregation with group by, Velox will pass all valid rows from all group in `addRawInput/addIntermediateResults`, it's hard and impossible to separate rows from different group, and we have no idea about how many groups need to process, so can not maintain a internal vector of values here.